docs/spec_draft.md

```diff
@@ -1267,6 +1268,132 @@ More formally, `result[i0, ..., iR-1]` is equal to:
+such that
+  * `jd = floor(((padding[d][0] + id * base_dilations[d]) + jd * (dilated_window[d] - window_strides[d])) /  window_dimensions[d]), where
+  dilated_window[d] = (window_dimensions[d] - 1) * window_dilations[d] + 1`
+  * `((padding[d] + id * base_dilations[d]) - jd * window_strides[d]) % window_dilations[d] = 0`. This is to skip the holes in dilated window.
+
+The evaluation order of the reduction function is arbitrary and may be
+non-deterministic. Therefore, the reduction function should not be overly
+sensitive to reassociation. See the discussion about associativity in the
+context of [stablehlo.reduce](#stablehloreduce) for more details.
+
+### Inputs
+
+| Name                | Type                                             | Constraints |
+|---------------------|--------------------------------------------------|-------------|
+| `inputs`            | variadic number of tensors of any supported type | (C1), (C3)  |
+| `init_values`       | variadic number of scalars of any supported type | (C1), (C2)  |
+| `window_dimensions` | 1-dimensional tensor constant type `si64`        | (C4)        |
+| `window_strides`    | 1-dimensional tensor constant type `si64`        | (C5)        |
+| `base_dilations`    | 1-dimensional tensor constant type `si64`        | (C6)        |
+| `window_dilations`  | 1-dimensional tensor constant type `si64`        | (C7)        |
+| `padding`           | 2-dimensional tensor constant type `si64`        | (C8)        |
+| `body`              | `function`                                       | (C9)        |
+
+### Outputs
+
+| Name     | Type                                              | Constraints |
+|----------|---------------------------------------------------|-------------|
+| `results` | variadic number of tensors of any supported type | (C1), (C8), (C9) |
+
+### Constraints
+
+  * (C1) size(`inputs`) $=$ size(`init_values`) $=$ size(`results`) $=$ N and
+         N $\ge$ 1
+
+  * (C2) All `init_values` values have to form an identity under reduction
+         `body`. The behavior of the operation is undefined otherwise.
+
+  * (C3) All `inputs` have the same shape.
+
+  * (C4) On `window_dimensions`
+    * size(`window_dimensions`) $=$ rank(`inputs[k]`) for any k $\in$ [0, N).
+    * `window_dimensions[i]` $\gt 0$ for all i $\in$
+      [0, size(`window_dimensions`))
+
+  * (C5) On `window_strides` (optional)
+    * Default value of 1 for each dimension of inputs[k] for any k $\in$ [0, N).
+    * size(`window_strides`) $=$ rank(`inputs[k]`) for any k $\in$ [0, N).
+    * `window_strides[i]` $\gt 0$ for all i $\in$ [0, size(`window_strides`))
+
+  * (C6) On `base_dilations` (optional)
+    * Default value of 1 for each dimension of inputs[k] for any k $\in$ [0, N).
+    * size(`base_dilations`) $=$ rank(`inputs[k]`) for any k $\in$ [0, N).
+    * `base_dilations[i]` $\gt 0$ for all i $\in$ [0, size(`base_dilations`))
+
+  * (C7) On `window_dilations` (optional)
```

Comment:
These are all related questions, but should we restrict the `window_dilations` to be up to shape of `inputs[k]` or leave it since a combination of `window_dimensions` and `window_strides` could affect the resulting window to be out of bounds anyway?

Reply:
Yes you are right. The dilated window could span beyond the shape of inputs[k] and if that happened the window is discarded for reduction computation.